src/pages/PortForwarding.vue

```diff
@@ -56,14 +85,67 @@ export default {
   },
 
   methods: {
-    onIncludeK8sServicesChanged(value) {
+    handleUpdatePort(newPort: number): void {
+      if (this.serviceBeingEdited) {
+        this.serviceBeingEdited.listenPort = newPort;
+      }
+    },
+
+    onIncludeK8sServicesChanged(value: boolean): void {
       if (value !== this.settings.portForwarding.includeKubernetesServices) {
         ipcRenderer.invoke('settings-write',
           { portForwarding: { includeKubernetesServices: value } } );
       }
     },
-  },
-};
+
+    compareServices(service1: K8s.ServiceEntry, service2: K8s.ServiceEntry): boolean {
+      return service1.name === service2.name &&
+        service1.namespace === service2.namespace &&
+        service1.port === service2.port;
+    },
+
+    findServiceMatching(serviceToMatch: K8s.ServiceEntry | undefined, serviceList: K8s.ServiceEntry[]): K8s.ServiceEntry | undefined {
+      if (!serviceToMatch) {
+        return undefined;
+      }
+      const compareServices = (service1: K8s.ServiceEntry, service2: K8s.ServiceEntry) => {
+        return service1.name === service2.name &&
+          service1.namespace === service2.namespace &&
+          service1.port === service2.port;
+      };
+
+      return serviceList.find(service => compareServices(service, serviceToMatch));
+    },
+
+    handleEditPortForward(service: K8s.ServiceEntry): void {
+      this.errorMessage = null;
+      if (this.serviceBeingEdited) {
+        ipcRenderer.invoke('service-forward', this.serviceBeingEdited, false);
+      }
+      this.serviceBeingEdited = Object.assign({}, service);
+      // Forward ServiceEntry without listenPort set to get random port.
+      // The user can change this after we get a random port.
+      ipcRenderer.invoke('service-forward', service, true);
+    },
+
+    handleCancelEditPortForward(service: K8s.ServiceEntry): void {
+      this.errorMessage = null;
+      ipcRenderer.invoke('service-forward', service, false);
```

Comment:
I think? This is one of the wrinkles that made this work difficult. When getting a random port, you have two options:

1. look ahead for a free port and then bind that in the next step, when it may have been taken
2. request any free port from the OS and let it pick the number for you

I chose 2. So once the user clicks `Forward`, the k8s port is already forwarded to a random port on the host. So, if they want to cancel the forwarding before entering their own port, we need to cancel the random forwarded port.

Reply:
Thanks, that makes sense to me 🙂